### PR TITLE
terminate

### PR DESCRIPTION
**Pull request by evilcc2018**

### CODE REVIEW

File: pai-experimental/cdp.sol [hunk 1]
```diff
@@ -7,16 +7,17 @@ pragma solidity 0.4.25;
...
-import "github.com/seeplayerone/dapp-bin/library/template.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/liquidator.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/price_oracle.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/pai_issuer.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/3rd/math.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/3rd/note.sol";
+import "github.com/evilcc2018/dapp-bin/library/template.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/liquidator.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/price_oracle.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/pai_issuer.sol";
```
Comment (jack-asimov) [inline pai-experimental/cdp.sol]: should change the import to 'seeplayerone' when merging back.

File: pai-experimental/cdp.sol [hunk 2]
```diff
@@ -314,8 +316,12 @@ contract CDP is DSMath, DSNote, Template {
 
     /// liquidate a CDP
     function liquidate(uint record) public note {
-        require(!safe(record) || settlement);
+        require(!safe(record));
+        require(!settlement);
```
Comment (jack-asimov) [inline pai-experimental/cdp.sol]: we can liquidate a CDP whenever the CDP is unsafe or we are in settlement stage.

File: pai-experimental/cdp.sol [hunk 3]
```diff
@@ -333,9 +339,26 @@ contract CDP is DSMath, DSNote, Template {
...
+        require(block.timestamp > closeTime + bufferPeriod );
+        uint upperLimit = min(liquidatedCDP + _num, CDPIndex);
+        for(uint i = liquidatedCDP + 1; i <= upperLimit; i++) {
+            if(CDPRecords[i].collateral > 0)
+                liquidateInternal(i);
+        }
+        liquidatedCDP = upperLimit;
+        if(liquidatedCDP == CDPIndex)
+            liquidator.settlePhaseTwo();
     }
```
Comment (jack-asimov) [inline pai-experimental/cdp.sol:358]: all calculations should use DSMath or SafeMath.

File: pai-experimental/cdp.sol [hunk 4]
```diff
@@ -348,5 +371,30 @@ contract CDP is DSMath, DSNote, Template {
...
+        return (CDPRecords[record].owner,CDPRecords[record].collateral,CDPRecords[record].accumulatedDebt1,CDPRecords[record].accumulatedDebt2);
+    }
+
+    function getFee() public view returns(uint){
+        return stabilityFee;
+    }
+
+    function getTax() public view returns(uint){
+        return governanceFee;
+    }
```
Comment (jack-asimov) [inline pai-experimental/cdp.sol]: to follow SAI conventions, fee -> governance fee and tax -> stability fee

File: pai-experimental/cdp.sol [hunk 5]
```diff
@@ -348,5 +371,30 @@ contract CDP is DSMath, DSNote, Template {
...
+    function checkSafe(uint record) public view returns (bool) {
+        return safe(record);
+    }
+
```
Comment (jack-asimov) [inline pai-experimental/cdp.sol:378]: safe() is not a view function

File: pai-experimental/liquidator.sol [hunk 6]
```diff
@@ -6,11 +6,11 @@ pragma solidity 0.4.25;
...
-import "github.com/seeplayerone/dapp-bin/pai-experimental/3rd/math.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/3rd/note.sol";
-import "github.com/seeplayerone/dapp-bin/library/template.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/pai_issuer.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/price_oracle.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/3rd/math.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/3rd/note.sol";
+import "github.com/evilcc2018/dapp-bin/library/template.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/pai_issuer.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/price_oracle.sol";
```
Comment (jack-asimov) [inline pai-experimental/liquidator.sol]: should be 'seeplayerone' in import.

File: pai-experimental/liquidator.sol [hunk 7]
```diff
@@ -118,9 +144,28 @@ contract Liquidator is DSMath, DSNote, Template {
...
+    }
+
+    function settlePhaseTwo() public {
+        require(settlement);
+        require(!allLiquidated);
+        ///下一语句需要改写为safemath的形式
+        if(flow.balance(this, ASSET_BTC) > 0)
+            collateralSettlementPrice = (totalDebt * (10**27) / flow.balance(this, ASSET_BTC));
+        allLiquidated = true;
+    }
```
Comment (jack-asimov) [inline pai-experimental/liquidator.sol:142]: should use DSMath

File: pai-experimental/pai_issuer.sol [hunk 8]
```diff
@@ -3,8 +3,8 @@ pragma solidity 0.4.25;
 // import "../library/template.sol";
 // import "./3rd/math.sol";
 
-import "github.com/seeplayerone/dapp-bin/library/template.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/3rd/math.sol";
+import "github.com/evilcc2018/dapp-bin/library/template.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/3rd/math.sol";
```
Comment (jack-asimov) [inline pai-experimental/pai_issuer.sol]: emmm...

File: pai-experimental/price_oracle.sol [hunk 9]
```diff
@@ -3,13 +3,14 @@ pragma solidity 0.4.25;
 // import "../library/template.sol";
 // import "./3rd/math.sol";
 
-import "github.com/seeplayerone/dapp-bin/library/template.sol";
-import "github.com/seeplayerone/dapp-bin/pai-experimental/3rd/math.sol";
+import "github.com/evilcc2018/dapp-bin/library/template.sol";
+import "github.com/evilcc2018/dapp-bin/pai-experimental/3rd/math.sol";
```
Comment (jack-asimov) [inline pai-experimental/price_oracle.sol]: emmm..

**Review by jack-asimov — CHANGES_REQUESTED** [review]
pls see the comments, thx.

File: pai-experimental/cdp.sol (hunk 1 above)
Comment (evilcc2018) [reply pai-experimental/cdp.sol]: OK

File: pai-experimental/cdp.sol (hunk 2 above)
Comment (evilcc2018) [reply pai-experimental/cdp.sol]: This need to be discussed. Of course you can make the final decision.

File: pai-experimental/cdp.sol (hunk 3 above)
Comment (evilcc2018) [reply pai-experimental/cdp.sol:358]: need your help

File: pai-experimental/cdp.sol (hunk 4 above)
Comment (evilcc2018) [reply pai-experimental/cdp.sol]: don't mind any method after '///only for debug'. I assume all these methods will be deleted after detailed testing.

File: pai-experimental/cdp.sol (hunk 5 above)
Comment (evilcc2018) [reply pai-experimental/cdp.sol:378]: same as last one

File: pai-experimental/liquidator.sol (hunk 7 above)
Comment (evilcc2018) [reply pai-experimental/liquidator.sol]: OK

Comment (evilcc2018) [reply pai-experimental/liquidator.sol:142]: need your help

File: pai-experimental/pai_issuer.sol (hunk 8 above)
Comment (evilcc2018) [reply pai-experimental/pai_issuer.sol]: OK

File: pai-experimental/price_oracle.sol (hunk 9 above)
Comment (evilcc2018) [reply pai-experimental/price_oracle.sol]: OK

**Comment by evilcc2018** [comment]
> pls see the comments, thx.

All finished.

**Pull request closed without merging**